compiler/rustc_builtin_macros/src/assert/context.rs

```diff
@@ -0,0 +1,81 @@
+use rustc_ast::{ptr::P, Expr, Path, Stmt};
+use rustc_data_structures::fx::FxHashSet;
+use rustc_expand::base::ExtCtxt;
+use rustc_span::{symbol::Ident, Span};
+
+pub(super) struct Context<'cx, 'a> {
+    // An optimization.
+    //
+    // Elements that aren't consumed (PartialEq, PartialOrd, ...) can be copied **after** the
+    // `assert!` expression fails rather than copied on-the-fly.
+    _best_case_captures: Vec<Stmt>,
+    // Top-level `let captureN = Capture::new()` statements
+    _capture_decls: Vec<Capture>,
+    cx: &'cx ExtCtxt<'a>,
+    // Top-level `let __local_bindN = &expr` statements
+    _local_bind_decls: Vec<Stmt>,
+    // Used to avoid capturing duplicated paths
+    //
+    // ```rust
+    // let a = 1i32;
+    // assert!(add(a, a) == 3);
+    // ```
+    _paths: FxHashSet<String>,
+    span: Span,
+}
+
+impl<'cx, 'a> Context<'cx, 'a> {
+    pub(super) fn new(cx: &'cx ExtCtxt<'a>, span: Span) -> Self {
+        Self {
+            _best_case_captures: <_>::default(),
+            _capture_decls: <_>::default(),
+            cx,
+            _local_bind_decls: <_>::default(),
+            _paths: <_>::default(),
```

Comment:
please remove the unused fields for now. Adding more fields to an existing data structure in the follow up PR won't make this harder to review.